### ISSUE TITLE
Hidden tunnel tests are broken

### ISSUE THREAD

**Issue by whirm** [Tribler/tribler#1826]
I'll disable the ones that fail all the time, but somebody should take a good look at them and fix whatever needs fixing. (see #1825)

**Comment by devos50** [comment]
@whirm maybe someone other than @lfdversluis and me can look at this one? Neither of us are experienced with the hidden tunnel community.

**Comment by whirm** [comment]
Maybe @Pathemeous?

@synctext ?

**Comment by Pathemeous** [comment]
I doubt that I have more practical knowledge of the matter than @lfdversluis or @devos50, but obviously I am also working in this area. Since I can't work on Tribler all week I cannot guarentee that I can fulfill the _Top Priority_ label, but I can make this the first thing I do when I do.

**Comment by whirm** [comment]
That would be great, It's top priority because it makes us have to disable those tests to be able to run the rest with a reasonable success rate, but as long as it gets fixed before any further work on the tunnel community so the community doesn't break further due to it not being properly tested, it should be fine.

**Comment by Pathemeous** [comment]
I think that should be doable. However, @synctext should probably confirm that that indeed no one is working on the tunnel community.

**Comment by whirm** [comment]
No one is as far as I know

**Comment by Pathemeous** [comment]
(y)

**Comment by whirm** [comment]
I'll assign this to you then

**Comment by synctext** [comment]
No one is working on the tunnel community...

**Comment by whirm** [comment]
@Pathemeous there's more tunnel issues on this same milestone you could help with

**Comment by Pathemeous** [comment]
When I'm on the seventh floor again I'll come to you, if that's all right

**Comment by devos50** [comment]
The tunnel community tests are fixed already, however, the hidden seeding test is still broken. We need to spent some time investigating the hidden tunnel community.

**Comment by qstokkink** [comment]
Small update on this, the `test_hidden_services` test is sprouting:
`Socks5Connection: ERROR: DENYING SOCKS5 request, reason: not enough circuits`
This means that, somehow, LibTorrent is trying to send before the TunnelCommunity has circuits at all (which -normally- is guarenteed by `can_create_engine_wrapper`).

**Comment by devos50** [comment]
@qstokkink we talked about that yesterday. Can you make it wait until the full e2e circuit is setup?

**Comment by qstokkink** [comment]
@devos50 I had both the seeder's and downloader's tunnel_community wait until they had a circuit and I still got that error, which _should_ be impossible.

**Comment by devos50** [comment]
@qstokkink is waiting until there's a circuit enough? I guess you could debug why it says that there are not enough circuits, even after waiting.

**Comment by qstokkink** [comment]
@devos50 normally this situation does not occur, but in the case of testing you need to make sure the TunnelCommunity has circuits before the Socks5 server receives data, otherwise you get that error (and you'll end up in an irrecoverable state).

My best guess is that the download manager, LaunchManyCore or something having to do with the downloader, has the wrong TunnelCommunity reference.

For now I'm mainly focused on thesis matters (2314, 2607), but if I make other notable discoveries I will update this thread.

**Comment by devos50** [comment]
Thanks. I don't think this issue will be solved before we release Tribler 7 so I will modify the milestone.

**Comment by qstokkink** [comment]
Small update on this: the linked e2e gets created correctly and the DHT-found peer is added to the download. However, upon inspection of this peer it is specified as follows:

```
[
   {
      'ip':'8.131.94.27',
      'dtotal':0,
      'downrate':0,
      'uinterested':False,
      'wstate':'\x00',
      'optimistic':False,
      'have':[
         False,
         False,
         False,
         False,
         False,
         False,
         False,
         False,
         False,
         False,
         False,
         False,
         False,
         False,
         False,
         False,
         False,
         False,
         False,
         False,
         False,
         False,
         False,
         False,
         False,
         False,
         False,
         False,
         False,
         False,
         False,
         False,
         False,
         False,
         False,
         False,
         False,
         False,
         False,
         False,
         False,
         False,
         False,
         False,
         False,
         False,
         False,
         False,
         False,
         False,
         False,
         False,
         False,
         False,
         False,
         False,
         False,
         False,
         False,
         False
      ],
      'id':'\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00',
      'speed':0,
      'port':1024,
      'connection_type':0,
      'extended_version':'',
      'dchoked':True,
      'utotal':0,
      'dinterested':False,
      'direction':'L',
      'uprate':0,
      'completed':0.0,
      'uchoked':True,
      'upload_only':False,
      'uhasqueries':False,
      'rstate':'\x00',
      'snubbed':False,
      'country':'\x00\x00',
      'uflushed':False,
      'seed':False
   }
]
```

So from the downloader's perspective the (anonymized) seeder has 0% file completion.

**Comment by qstokkink** [comment]
@egbertbouman In this experiment the SOCKS5 server is firing off a [`cannot send data, no clue where to send it to`](https://github.com/Tribler/tribler/blob/devel/Tribler/community/tunnel/Socks5/server.py#L45) for the seeder (and only the seeder, the downloader's SOCKS5 server works fine). Furthermore, I am seeing [`UDP_ASSOCIATE` requests coming in for association with `('0.0.0.0', 0)`](https://github.com/Tribler/tribler/blob/devel/Tribler/community/tunnel/Socks5/server.py#L227) on the seeder side.
Do you have an idea about what is going on here?

**Comment by egbertbouman** [comment]
@qstokkink  If I remember correctly there used to be some logic at the seeder side that ensured libtorrent made an outgoing connection. I'll have a look and create a PR to see if this fixes it...

**Comment by qstokkink** [comment]
@egbertbouman thanks! I think this is (close to) the final piece of the puzzle to get this test working again.

**Comment by qstokkink** [comment]
Resolved by #2792.

**Comment by devos50** [comment]
Finally 🎉
